osu.Game/Online/Chat/MessageNotifier.cs

```diff
@@ -0,0 +1,234 @@
+using osu.Game.Users;
+
+namespace osu.Game.Online.Chat
+{
+    /// <summary>
+    /// Component that handles creating and posting notifications for incoming messages.
+    /// </summary>
+    public class MessageNotifier : Component
+    {
+        [Resolved(CanBeNull = true)]
+        private NotificationOverlay notificationOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChatOverlay chatOverlay { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        private ChannelManager channelManager { get; set; }
+
+        private Bindable<bool> notifyOnMention;
+        private Bindable<bool> notifyOnChat;
+        private Bindable<User> localUser;
+
+        /// <summary>
+        /// Determines if the user is able to see incoming messages.
+        /// </summary>
+        public bool IsActive => chatOverlay?.IsPresent == true;
+
+        private readonly List<PrivateMessageNotification> privateMessageNotifications = new List<PrivateMessageNotification>();
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config, IAPIProvider api)
+        {
+            notifyOnMention = config.GetBindable<bool>(OsuSetting.ChatHighlightName);
+            notifyOnChat = config.GetBindable<bool>(OsuSetting.ChatMessageNotification);
+            localUser = api.LocalUser;
+
+            // Listen for new messages
+            channelManager.JoinedChannels.ItemsAdded += joinedChannels =>
+            {
+                foreach (var channel in joinedChannels)
+                    channel.NewMessagesArrived += channel_NewMessagesArrived;
+            };
+
+            channelManager.JoinedChannels.ItemsRemoved += leftChannels =>
+            {
+                foreach (var channel in leftChannels)
+                    channel.NewMessagesArrived -= channel_NewMessagesArrived;
+            };
+        }
+
+        private void channel_NewMessagesArrived(IEnumerable<Message> messages)
+        {
+            if (messages == null || !messages.Any())
+                return;
+
+            HandleMessages(messages.First().ChannelId, messages);
+        }
+
+        /// <remarks>
+        /// Resolves the channel id
+        /// </remarks>
+        public void HandleMessages(long channelId, IEnumerable<Message> messages)
+        {
+            var channel = channelManager.JoinedChannels.FirstOrDefault(c => c.Id == channelId);
+
+            if (channel == null)
+            {
+                Logger.Log($"Couldn't resolve channel id {channelId}", LoggingTarget.Information);
+                return;
+            }
+
+            HandleMessages(channel, messages);
+        }
+
+        public void HandleMessages(Channel channel, IEnumerable<Message> messages)
+        {
+            // don't show if the ChatOverlay and the target channel is visible.
+            if (IsActive && channelManager.CurrentChannel.Value == channel)
+                return;
+
+            foreach (var message in messages)
```

Comment:
After reading this loop it appears clear to me that this structure isn't working for this usecase, there's too much branching. This would benefit in inversion of control via a message processor abstraction. Consider the following:

* Create N message processors (one for mentions, one for PMs, ...)
* Processors are applicable to certain channel types (mentions - global, PMs - private channels only)
* Processors take in messages and do whatever they need to with them

This yields:

* Less branching
* Better separation of concerns
* They're testable!
* You could even hide the lists of notifications sent lower down to the processors making them a little less painful

I haven't tried to implement, but that's the way I'd try to go with this code. At current it is really difficult to keep track of all of this.

Reply:
The problem is, even if I would take this effort to make this all objects with properties and what not. This might or will get deprecated once web notifications get integrated into the game (like peppy warned me before hand), **unless** we have a place/use case for such system.

Reply:
While you're correct, I don't think this is in a mergeable state at current, even if this PR is temporary. However I'm not going to insist - there is already improvement on the original implementation as far as I'm concerned - I'll defer to the judgement of others.